### PR TITLE
 feat: allow redirect to external url

### LINKED ISSUE #770: Allow redirect to absolute URL
Feature request: Allow redirects to absolute URLs, SSR and client-side.

We have an app (https://domainr.com) with several legacy URLs, some of which redirect to other websites. We’re exploring using Nuxt/Vue for our new website and would like to preserve this behavior.

## Snippet to reproduce

`~/pages/_domain/iana.vue`

```javascript
<script>
export default {
  fetch ({ params, redirect }) {
    redirect(301, `https://www.iana.org/domains/root/db/${params.domain}.html`)
  }
}
</script>
```

## Steps to reproduce

Navigate to `/com/iana`

## Expected

Redirect to `https://www.iana.org/domains/root/db/com.html`

(Click here to see expected behavior: https://domainr.com/com/iana)

## What actually happens

Redirect to `/https:/www.iana.org/domains/root/db/com.html` (local URL)

## Background

Nuxt only allows local redirects to a path, which is prepended with `router.base` here: https://github.com/nuxt/nuxt.js/blob/dev/lib/app/server.js#L39

### PR DESCRIPTION
**Pull request by clarkdo**
Resolve #770 #2287, refactor  #2055 
- [x] add formatUrl method to format `url` with `query object`
- [x] allow redirect to external link, such as:
    1. redirect('github.com/nuxt/nuxt.js')
    2. redirect('https://example.com/api/data?query=test', {query2: 'test2'})
- [x] add test

### CODE REVIEW

File: lib/app/utils.js [hunk 1]
```diff
@@ -132,11 +132,23 @@ export async function setContext(app, context) {
         path = status
         status = 302
       }
-      app.context.next({
-        path: path,
-        query: query,
-        status: status
-      })
+      if (path.charAt(0) === '/' && path.charAt(1) !== '/') {
```
Comment (pi0) [inline lib/app/utils.js]: Does this means we don't support relative URLs anymore? Something like `./comments` will be an external URL :)

File: lib/app/components/nuxt.js [hunk 2]
```diff
@@ -7,6 +7,9 @@ export default {
   name: 'nuxt',
   props: ['nuxtChildKey'],
   render(h) {
+    if (this.nuxt._redirected) {
+      return h('div', [ 'redirecting.' ])
```
Comment (pi0) [inline lib/app/components/nuxt.js]: Maybe we can support for i18n of this message :) I will handle it

**Review by pi0 — COMMENTED** [review]
.

File: lib/app/components/nuxt.js (hunk 2 above)
Comment (clarkdo) [reply lib/app/components/nuxt.js]: What about of  moving it to `options.messages` ? If ok, I will open a pr 😄

File: lib/app/utils.js (hunk 1 above)
Comment (clarkdo) [reply lib/app/utils.js]: @pi0 Yes, I just read the `vue-router` sources, I will make a change.

File: lib/app/components/nuxt.js (hunk 2 above)
Comment (pi0) [reply lib/app/components/nuxt.js]: If you can do it it would be awesome 😍

File: lib/app/utils.js [hunk 3]
```diff
@@ -132,11 +132,23 @@ export async function setContext(app, context) {
...
+        })
+      } else {
+        path = formatUrl(path, query)
+        if (process.server) {
+          app.context.res.setHeader('Location', path)
+          app.context.res.statusCode = status
+          app.nuxt._redirected = true
+        }
+        if (process.client) {
+          window.location = path
```
Comment (ydnar) [inline lib/app/utils.js]: This should be `window.location.replace(path)`. The current form adds an additional history item.

Reproduce by clicking a redirect link, then hit Back twice. It should go back after one click.

Comment (pi0) [reply lib/app/utils.js]: Already fixed in dev branch :)

Comment (Atinux) [reply lib/app/utils.js]: Thank you @ydnar

Comment (ydnar) [reply lib/app/utils.js]: Sweet as!

**Pull request merged**